hudi-client/hudi-client-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadataWriter.java

```diff
@@ -1097,9 +1097,9 @@ engineContext, dataWriteConfig, commitMetadata, instantTime, dataMetaClient, get
 
       // Updates for record index are created by parsing the WriteStatus which is a hudi-client object. Hence, we cannot yet move this code
       // to the HoodieTableMetadataUtil class in hudi-common.
-      if (dataWriteConfig.isRecordIndexEnabled()) {
-        HoodieData<HoodieRecord> additionalUpdates = getRecordIndexAdditionalUpserts(partitionToRecordMap.get(MetadataPartitionType.RECORD_INDEX.getPartitionPath()), commitMetadata);
-        partitionToRecordMap.put(RECORD_INDEX.getPartitionPath(), partitionToRecordMap.get(MetadataPartitionType.RECORD_INDEX.getPartitionPath()).union(additionalUpdates));
+      if (dataWriteConfig.isRecordIndexEnabled() && RECORD_INDEX.isMetadataPartitionAvailable(dataMetaClient)) {
```

Comment:
guess this was not the design. 
the regular ingestion writer should log updates for partitions fully available and for partitions that are inflight. 
if not, the async indexer has to do full catch up. The catch up is meant just for a race condition to ensure we do not miss any updates, but in most cases, the catch up phase should not have any work to do. 
lets discuss if our understanding is different

Reply:
As discussed, I have changed the condition here.